cmd/gateway_server/server.go

```diff
@@ -0,0 +1,190 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"strings"
+
+	"github.com/WICG/webpackage/go/signedexchange"
+	"github.com/WICG/webpackage/go/signedexchange/certurl"
+	pb "github.com/ampproject/amppackager/cmd/gateway_server/gateway"
+	"github.com/ampproject/amppackager/packager/rtv"
+	"github.com/ampproject/amppackager/packager/signer"
+	"github.com/ampproject/amppackager/packager/util"
+	"github.com/julienschmidt/httprouter"
+	"google.golang.org/grpc"
+)
+
+var (
+	port                = flag.Int("port", 9000, "Gateway server port")
+	publisherServerPort = flag.Int("publisher_server_port", 10000,
+		"Publisher server port.")
+	privateKey = flag.String("private_key", "",
+		"Path to private key. Must be same private key used to generate certs.")
+)
+
+type gatewayServer struct{}
+
+func shouldPackage() bool {
+	return true
+}
+
+func errorToSXGResponse(err error) *pb.SXGResponse {
+	response := &pb.SXGResponse{
+		Error:            true,
+		ErrorDescription: err.Error(),
+	}
+	return response
+}
+
+func (s *gatewayServer) GenerateSXG(ctx context.Context, request *pb.SXGRequest) (*pb.SXGResponse, error) {
+	rtvCache, err := rtv.New()
+	if err != nil {
+		return errorToSXGResponse(err), nil
+	}
+
+	certs, err := signedexchange.ParseCertificates(request.PublicCert)
+	if err != nil {
+		return errorToSXGResponse(err), nil
+	}
+
+	privateKey, err := util.ParsePrivateKey(request.PrivateKey)
+	if err != nil {
+		return errorToSXGResponse(err), nil
+	}
+
+	signUrl, err := url.Parse(request.SignUrl)
+	if err != nil {
+		return errorToSXGResponse(err), nil
+	}
+
+	var dotStarPattern = ".*"
+	signUrlPattern := util.URLPattern{
+		Domain:  signUrl.Host,
+		QueryRE: &dotStarPattern,
+	}
+	err = util.ValidateSignURLPattern(&signUrlPattern)
+	if err != nil {
+		return errorToSXGResponse(err), nil
+	}
+
+	fetchUrlPattern := util.URLPattern{
+		Scheme:                 []string{"http"},
+		Domain:                 fmt.Sprintf("localhost:%d", *publisherServerPort),
+		ErrorOnStatefulHeaders: false,
+		QueryRE:                &dotStarPattern,
+		SamePath:               new(bool),
+	}
+	*fetchUrlPattern.SamePath = false
+	err = util.ValidateFetchURLPattern(&fetchUrlPattern)
+	if err != nil {
+		return errorToSXGResponse(err), nil
+	}
+
+	urlSets := []util.URLSet{
+		{
+			Sign:  &signUrlPattern,
+			Fetch: &fetchUrlPattern,
+		},
+	}
+
+	packager, err := signer.New(certs[0], privateKey, urlSets, rtvCache, shouldPackage, signUrl, false)
+
+	if err != nil {
+		return errorToSXGResponse(err), nil
+	}
+
+	if packager == nil {
+		return errorToSXGResponse(err), nil
```

Comment:
But if you do keep it around, replace `err` here with a hard-coded string, for a more useful SXGResponse.